File: src/main/java/fr/maxlego08/satisfactorydle/GuildConfigManager.java
package fr.maxlego08.satisfactorydle;

import fr.maxlego08.sarah.RequestHelper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class GuildConfigManager {

    private static final String TABLE = "guild_configs";
    private static final String ALL_MODES = String.join(",",
            Arrays.stream(GameMode.values()).map(GameMode::getKey).toArray(String[]::new));

    private final RequestHelper requestHelper;
    private final String defaultLocale;
    private final Map<String, GuildConfig> cache = new ConcurrentHashMap<>();

    public GuildConfigManager(RequestHelper requestHelper, String defaultLocale) {
        this.requestHelper = requestHelper;
        this.defaultLocale = defaultLocale;
        loadAll();
    }

    private void loadAll() {
        List<GuildConfig> configs = requestHelper.selectAll(TABLE, GuildConfig.class);
        for (GuildConfig config : configs) {
            cache.put(config.getGuildId(), config);
        }
    }

    public GuildConfig getConfig(String guildId) {
        return cache.computeIfAbsent(guildId, id -> {
            GuildConfig config = new GuildConfig(id, defaultLocale, ALL_MODES);
            save(config);
            return config;
        });
    }

    public void setLocale(String guildId, String locale) {
        GuildConfig current = getConfig(guildId);
        GuildConfig updated = new GuildConfig(guildId, locale, current.getActiveModes());
        cache.put(guildId, updated);
        save(updated);
    }

    public void setModeEnabled(String guildId, String mode, boolean enabled) {
        GuildConfig current = getConfig(guildId);
        List<String> modes = new ArrayList<>(current.getActiveModeList());
        if (enabled && !modes.contains(mode)) {
            modes.add(mode);
        } else if (!enabled) {
            modes.remove(mode);
        }
        String activeModes = String.join(",", modes);
        GuildConfig updated = new GuildConfig(guildId, current.getLocale(), activeModes);
        cache.put(guildId, updated);
        save(updated);
    }

    private void save(GuildConfig config) {
        requestHelper.upsert(TABLE, GuildConfig.class, config);
    }
}

File: app/storage/guild_config_manager.py
from __future__ import annotations

import aiosqlite

from app.domain.game_mode import GameMode
from app.domain.guild_config import GuildConfig


class GuildConfigManager:
    ALL_MODES = ",".join(mode.value for mode in GameMode)

    def __init__(self, sqlite_path: str, default_locale: str) -> None:
        self.sqlite_path = sqlite_path
        self.default_locale = default_locale
        self._cache: dict[str, GuildConfig] = {}

    async def load_all(self) -> None:
        async with aiosqlite.connect(self.sqlite_path) as db:
            async with db.execute("SELECT guild_id, locale, active_modes FROM guild_configs") as cursor:
                rows = await cursor.fetchall()

        for guild_id, locale, active_modes in rows:
            self._cache[guild_id] = GuildConfig(guild_id=guild_id, locale=locale, active_modes=active_modes)

    async def get_config(self, guild_id: str) -> GuildConfig:
        cached = self._cache.get(guild_id)
        if cached is not None:
            return cached

        config = GuildConfig(guild_id=guild_id, locale=self.default_locale, active_modes=self.ALL_MODES)
        self._cache[guild_id] = config
        await self._save(config)
        return config

    async def set_locale(self, guild_id: str, locale: str) -> None:
        current = await self.get_config(guild_id)
        updated = GuildConfig(guild_id=guild_id, locale=locale, active_modes=current.active_modes)
        self._cache[guild_id] = updated
        await self._save(updated)

    async def set_mode_enabled(self, guild_id: str, mode: str, enabled: bool) -> None:
        current = await self.get_config(guild_id)
        modes = current.get_active_mode_list()
        if enabled and mode not in modes:
            modes.append(mode)
        elif not enabled:
            modes = [value for value in modes if value != mode]

        updated = GuildConfig(guild_id=guild_id, locale=current.locale, active_modes=",".join(modes))
        self._cache[guild_id] = updated
        await self._save(updated)

    async def _save(self, config: GuildConfig) -> None:
        query = """
        INSERT INTO guild_configs(guild_id, locale, active_modes)
        VALUES (?, ?, ?)
        ON CONFLICT(guild_id) DO UPDATE SET
            locale = excluded.locale,
            active_modes = excluded.active_modes
        """
        async with aiosqlite.connect(self.sqlite_path) as db:
            await db.execute(query, (config.guild_id, config.locale, config.active_modes))
            await db.commit()
